### PR TITLE
Update wire order support for probabilities

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** This PR introduces direct support for out-of-order wire indices with LightningQubit by providing permutation of the probabilities tensor using Jet's `DefaultPermuter` class.

TODO: Add Jet and associated source origin annotations before merging.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mlxd** [comment]
[sc-62467]

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        PL_ABORT_IF_NOT(shape.size() == new_order.size(),
+                        "Tensor shape does not match number of indices.");
+        PL_ABORT_IF_NOT(data_size == data_in.size(),
+                        "Tensor shape does not match given input tensor data.");
+        PL_ABORT_IF_NOT(
+            data_size == data_out.size(),
+            "Tensor shape does not match given output tensor data.");
+        PL_ABORT_IF_NOT(
+            idx_old == idx_new,
+            "New indices are an invalid permutation of the existing indices");
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:74]:
```suggestion
            "New indices are an invalid permutation of the existing indices.");
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        PL_ABORT_IF_NOT(idx_new.size() == new_order.size(),
+                        "Duplicate transpose indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(shape.size() == new_order.size(),
+                        "Tensor shape does not match number of indices.");
+        PL_ABORT_IF_NOT(data_size == data_in.size(),
+                        "Tensor shape does not match given tensor data.");
+        PL_ABORT_IF_NOT(
+            idx_old == idx_new,
+            "New indices are an invalid permutation of the existing indices");
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:113]:
```suggestion
            "New indices are an invalid permutation of the existing indices.");
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+        PL_ABORT_IF(shape.empty(), "Tensor shape cannot be empty.");
+        PL_ABORT_IF(new_order.empty(), "Tensor indices cannot be empty.");
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:116]: Do we need to add these two checks to the other `Transpose` method? You could use a private helper method for these checks to remove the duplication of these checks in both methods.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  protected:
+    friend PermuterBackend;
+
+  private:
+    PermuterBackend permuter_b_;
+};
+
+/**
+ * @brief Default Permuter backend class for generalised transforms. Adapted
+ * from QFlex.
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:129]: Would you like to add the link to QFlex code here?

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+ */
+template <size_t BLOCKSIZE = 1024> class DefaultPermuter {
+
+  public:
+    /**
+     * @brief Reference-based transpose operation. See `Permuter` class for more
+     * details.
+     */
+    template <class T>
+    void Transpose(const std::vector<T> &data_,
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:141]: did you mean `data_in`?

```suggestion
    void Transpose(const std::vector<T> &data_in,
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+                   const std::vector<size_t> &shape, std::vector<T> &data_out,
+                   const std::vector<std::string> &old_indices,
+                   const std::vector<std::string> &new_indices) {
+        data_out = data_;
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:145]:
```suggestion
        data_out = data_in;
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+        if (new_indices == old_indices)
+            return;
+
+        const std::size_t num_indices = old_indices.size();
+        const std::size_t total_dim = data_.size();
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:151]:
```suggestion
        const std::size_t total_dim = data_in.size();
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+        // offset is important when doing this in blocks, as it's indeed
+        // implemented.
+        std::size_t offset = 0;
+        // internal_po keeps track of interations within a block.
+        // Blocks have size `blocksize`.
+        std::size_t internal_po = 0;
+
+        T *data = data_out.data();
+        const T *scratch =
+            data_.data(); // internal pointer offers better performance than
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:198]:
```suggestion
            data_in.data(); // internal pointer offers better performance than
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+            remaining_data -= blocksize_;
+        }
+    }
+
+    /**
+     * @brief Return-based transpose operation. See `Permuter` class for more
+     * details.
+     */
+    template <class T>
+    std::vector<T> Transpose(std::vector<T> data_,
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:255]:
```suggestion
    std::vector<T> Transpose(std::vector<T> data_in,
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+                             const std::vector<size_t> &shape,
+                             const std::vector<std::string> &old_indices,
+                             const std::vector<std::string> &new_indices) {
+        std::vector<T> data_out(std::move(data_));
+        Transpose(data_, shape, data_out, old_indices, new_indices);
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:260]:
```suggestion
        std::vector<T> data_out(std::move(data_in));
        Transpose(data_in, shape, data_out, old_indices, new_indices);
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        if (new_indices == old_indices)
+            return;
+
+        const std::size_t num_indices = old_indices.size();
+        const std::size_t total_dim = data_.size();
+        std::size_t remaining_data = total_dim;
+
+        if (num_indices == 0) {
+            PL_ABORT("Number of indices cannot be zero.");
+        }
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:156]:
```suggestion
        PL_ABORT_IF_NOT(num_indices, "Number of indices cannot be zero.");
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+
+        // Create map_old_to_new_idxpos from old to new indices, and
+        // new_dimensions.
+        std::vector<size_t> map_old_to_new_idxpos(num_indices);
+        std::vector<size_t> new_dimensions(num_indices);
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:161]:
```suggestion
        std::vector<std::size_t> map_old_to_new_idxpos(num_indices);
        std::vector<std::size_t> new_dimensions(num_indices);
```

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+        for (size_t i = 0; i < num_indices; ++i) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:162]: would you like to use OMP here?

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+            // Copy data for this block, taking into account offset of
+            // small_map...
+            effective_max = std::min(blocksize_, remaining_data);
+            for (size_t p = 0; p < effective_max; p++) {
+                data[small_map_old_to_new_position[p]] = scratch[offset + p];
+            }
+
+            offset += blocksize_;
+            remaining_data -= blocksize_;
+        }
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/NDPermuter.hpp:247]: nice!

File: pennylane_lightning/core/src/utils/tests/Test_NDPermuter.cpp [hunk 15]
```diff
@@ -0,0 +1,35 @@
+
+// Copyright 2018-2023 Xanadu Quantum Technologies Inc.
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/tests/Test_NDPermuter.cpp:2]:
```suggestion
// Copyright 2024 Xanadu Quantum Technologies Inc.
```

File: pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp [hunk 16]
```diff
@@ -125,12 +128,29 @@ class Measurements final
...
             ind_probs++;
         }
-        // Transposing the probabilities tensor with the indices determined
-        // at the beginning.
+
+        // Permute the data according to the required wire ordering
         if (wires != sorted_wires) {
-            probabilities = Pennylane::Util::transpose_state_tensor(
-                probabilities, sorted_ind_wires);
+            static constexpr std::size_t CACHE_SIZE = 8;
```
Comment (maliasadi) [inline pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:134]: What's the reason behind this magic cache size? It's a little bit two low.

Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:134]: It's fine for now. This is from JET, and used to ensure reasonable partition across the axes. Too high, and wasteful, too low, and not optimal, but functional. We can optimise on this later once the release is all sorted.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp [hunk 17]
```diff
@@ -226,25 +226,30 @@ TEMPLATE_TEST_CASE("Probabilities", "[Measures]", float, double) {
...
+        // 0.00397696,
+        //  0.00801973, 0.00104134}},
+        //{{2, 0, 1},
+        // {0.67078706, 0.17564072, 0.03062806, 0.00801973, 0.0870997,
+        // 0.02280642,
+        //  0.00397696, 0.00104134}},
+        //{{2, 1, 0},
+        // {0.67078706, 0.17564072, 0.0870997, 0.02280642, 0.03062806,
+        // 0.00801973,
+        //  0.00397696, 0.00104134}}, // TODO: Fix LK out-of-order permutations
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp:248]: Are we bringing these back?

Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp:248]: Not for this release, for now we raise an error. These are wrong, as with the LQ test results for the same ordering, so we will need to fix LK's transpose too later. TODO in there will be enough to track

File: pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp [hunk 18]
```diff
@@ -226,25 +226,30 @@ TEMPLATE_TEST_CASE("Probabilities", "[Measures]", float, double) {
...
         {{0, 1}, {0.70141512, 0.09107666, 0.18366045, 0.02384776}},
         {{0, 2}, {0.75788676, 0.03460502, 0.19844714, 0.00906107}},
         {{1, 2}, {0.84642778, 0.0386478, 0.10990612, 0.0050183}},
-        {{2, 1}, {0.84642778, 0.10990612, 0.0386478, 0.0050183}},
+        //{{2, 1}, {0.84642778, 0.10990612, 0.0386478, 0.0050183}},
```
[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/tests/Test_StateVectorKokkos_Measure.cpp:252]: Ok, will remove in the next push

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 13 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:162]: I think we should avoid any optimisation of the direct-from-JET functions now. If it is working, we can remove it for the multi-threading version later, which requires more integration work.

File: .github/workflows/tests_linux_cpp.yml [hunk 19]
```diff
@@ -132,6 +152,27 @@ jobs:
     runs-on: ${{ matrix.os }}
 
     steps:
+      - name: Remove GH Action tooling
```
[the remark this answers is not in the document]
Comment (mlxd) [reply .github/workflows/tests_linux_cpp.yml:158]: given today's time pressure for release let's worry about this later

[the remark this answers is not in the document]
Comment (vincentmr) [reply .github/workflows/tests_linux_cpp.yml:158]: Or create a workflow and perform a workflow call?

**Comment by mlxd** [comment]
Hey all, I'll be rejecting all suggestions to the NDPermuter class, unless they improve correctness of results. This is to ensure we can actually get everything ready for release today.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 11 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:156]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 10 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:260]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 9 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:255]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 8 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:198]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 7 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:151]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 6 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:145]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 5 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:141]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 4 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:129]: I'll link to JET once we are done, which handles the attribution notices

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 3 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:116]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 1 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:74]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp [hunk 20]
```diff
@@ -0,0 +1,268 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:1]: This notice will need to match JET. Will resolve once we are done with correctness of results.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 12 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:161]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/utils/NDPermuter.hpp (hunk 13 above)
Comment (mlxd) [reply pennylane_lightning/core/src/utils/NDPermuter.hpp:162]: To be resolved by later importing directly from JET. Resolving.

File: pennylane_lightning/core/src/simulators/lightning_qubit/measurements/tests/Test_MeasurementsLQubit.cpp [hunk 21]
```diff
@@ -81,17 +83,17 @@ TEMPLATE_PRODUCT_TEST_CASE("Expected Values", "[Measurements]",
         operations_list = {PauliX, PauliX, PauliX};
         exp_values = Measurer.expval(operations_list, wires_list);
         exp_values_ref = {0.49272486, 0.42073549, 0.28232124};
-        REQUIRE_THAT(exp_values, Catch::Approx(exp_values_ref).margin(1e-6));
+        CHECK_THAT(exp_values, Catch::Approx(exp_values_ref).margin(1e-6));
```
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_qubit/measurements/tests/Test_MeasurementsLQubit.cpp:86]: Why the changes to `CHECK_THAT`?

Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_qubit/measurements/tests/Test_MeasurementsLQubit.cpp:86]: Because first fail doesn't give us enough info to understand the cause. CHECK give more details as the failures don't stop the process.

File: pennylane_lightning/core/src/utils/tests/Test_NDPermuter.cpp [hunk 22]
```diff
@@ -0,0 +1,35 @@
+
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/utils/tests/Test_NDPermuter.cpp:1]:
```suggestion
```

**Review by AmintorDusko — APPROVED** [review]
Nice job!

**Review by maliasadi — APPROVED** [review]
Happy to merge this PR asap! :100: We can resolve the remaining suggestions in a followup PR after the release.

**Comment by mlxd** [comment]
Thanks for the promptness with this folks
